### PR TITLE
Enhance list-checks command

### PR DESCRIPTION
**Pull request by ernilambar**
Refactor list-checks command with improvements.

* Fix: `get_checks()` of `Default_Check_Repository` by default does not return experimental checks. This PR fix that issue for returing all checks with parameter `Default_Check_Repository::TYPE_ALL | Default_Check_Repository::INCLUDE_EXPERIMENTAL`.
* Addition: `--categories` argument is added to filter checks by categories.
* Addition: `--stability` argument is added to filter checks by stability. Takes either `stable` or `experimental`.

Output:

`wp plugin list-checks --stability=stable` - Currently all checks will be displayed as we dont have experimental checks yet.

stability

`wp plugin list-checks --categories=performance`

single-category

`wp plugin list-checks --categories=performance,general`

multiple-category

### CODE REVIEW

**Comment by ernilambar** [comment]
cc: @swissspidy @felixarntz @mukeshpanchal27

**Review by swissspidy — APPROVED** [review]

File: includes/CLI/Plugin_Check_Command.php [hunk 1]
```diff
@@ -266,18 +277,25 @@ public function list_checks( $args, $assoc_args ) {
 
 		$all_checks = array();
 
-		foreach ( $check_repo->get_checks() as $key => $check ) {
+		foreach ( $check_repo->get_checks( Default_Check_Repository::TYPE_ALL | Default_Check_Repository::INCLUDE_EXPERIMENTAL ) as $key => $check ) {
```
Comment (swissspidy) [inline includes/CLI/Plugin_Check_Command.php]: These constants or on the `Check_Repository` class, not `Default_Check_Repository`.

Also, can we do the filtering here instead? The checks collection returned by `get_checks()` has a `filter` method for this purpose.

```suggestion
		foreach ( $check_repo->get_checks( Check_Repository::TYPE_ALL | Check_Repository::INCLUDE_EXPERIMENTAL )->filter(
			static function ( Check $check ) {
				// Filter by $check->get_categories() and $check->get_stability()
				// return bool
			}
		) as $key => $check ) {
```

File: tests/behat/features/plugin-list-checks.feature [hunk 2]
```diff
@@ -14,3 +14,32 @@ Feature: Test that the WP-CLI plugin list checks command works.
...
+      """
+
+    When I run the WP-CLI command `plugin list-checks --format=json --stability=experimental`
+    Then STDOUT should be:
+      """
+      []
+      """
+
+    When I run the WP-CLI command `plugin list-checks --format=csv --categories=general`
+    Then STDOUT should be:
```
Comment (mukeshpanchal27) [inline tests/behat/features/plugin-list-checks.feature]:
```suggestion
    Then STDOUT should contain:
```

Given the current situation, this test will pass. However, when new checks are added, it is anticipated to fail. Please modify it to pass only if it contains certain criteria.

File: tests/behat/features/plugin-list-checks.feature [hunk 3]
```diff
@@ -14,3 +14,32 @@ Feature: Test that the WP-CLI plugin list checks command works.
...
+      """
+      slug,category,stability
+      i18n_usage,general,stable
+      """
+
+    When I run the WP-CLI command `plugin list-checks --format=csv --categories="general, security"`
+    Then STDOUT should be:
```
Comment (mukeshpanchal27) [inline tests/behat/features/plugin-list-checks.feature]: Check similar feedback comment.

File: includes/CLI/Plugin_Check_Command.php [hunk 4]
```diff
@@ -289,6 +307,35 @@ public function list_checks( $args, $assoc_args ) {
...
+			);
+		}
+
+		// Display categories in better way.
+		array_walk(
+			$all_checks,
+			static function ( &$item ) {
+				$item['category'] = join( ', ', $item['category'] );
+			}
+		);
```
Comment (mukeshpanchal27) [inline includes/CLI/Plugin_Check_Command.php]: Could you please explain why the changes needed?

**Review by mukeshpanchal27 — CHANGES_REQUESTED** [review]

File: includes/CLI/Plugin_Check_Command.php (hunk 4 above)
Comment (ernilambar) [reply includes/CLI/Plugin_Check_Command.php]: WP-CLI displays array value like this `["General", "Performance"]`. So we are converting to display like `general, performance`.

File: includes/CLI/Plugin_Check_Command.php (hunk 1 above)
Comment (felixarntz) [reply includes/CLI/Plugin_Check_Command.php]: I think retrieving all checks here is not the right approach. IMO it should behave the same way as `wp plugin check` does:
* By default, only stable checks are returned.
* Instead of a `--stability` option, we should use a boolean flag option `--include-experimental`.
* Only if that flag is set, we should get both stable and experimental checks.

File: includes/CLI/Plugin_Check_Command.php [hunk 5]
```diff
@@ -289,6 +307,35 @@ public function list_checks( $args, $assoc_args ) {
...
 
+		// Filter by stability.
+		if ( ! empty( $options['stability'] ) ) {
+			$all_checks = array_filter(
+				$all_checks,
+				static function ( $item ) use ( $options ) {
+					return $options['stability'] === $item['stability'];
+				}
+			);
+		}
```
Comment (felixarntz) [inline includes/CLI/Plugin_Check_Command.php]: See above: This shouldn't be needed as we should instead only retrieve experimental checks in the first place if the flag is set.

File: includes/CLI/Plugin_Check_Command.php [hunk 6]
```diff
@@ -289,6 +307,35 @@ public function list_checks( $args, $assoc_args ) {
...
+
+		// Filter by categories.
+		if ( ! empty( $options['categories'] ) ) {
+			$all_checks = array_filter(
+				$all_checks,
+				static function ( $item ) use ( $options ) {
+					$categories = array_map( 'trim', explode( ',', $options['categories'] ) );
+					return array_intersect( $categories, $item['category'] );
+				}
+			);
```
Comment (felixarntz) [inline includes/CLI/Plugin_Check_Command.php]: This should be simplified by reusing `Check_Categories::filter_checks_by_categories()` which is already implemented for this purpose.

File: includes/CLI/Plugin_Check_Command.php (hunk 4 above)
Comment (felixarntz) [reply includes/CLI/Plugin_Check_Command.php]: I don't think this is needed. Before you had the `join()` call when creating the `$all_checks` list, and I think it can be kept that way.

The only thing you'll have to make sure of is that you do all the check filtering using the standard methods on the original check collection, _before_ creating the `$all_checks` variable for CLI output.

**Review by felixarntz — CHANGES_REQUESTED** [review]
@ernilambar I think the ideas here make sense, however the implementation can be simplified.

**Comment by ernilambar** [comment]
Updates:

* `get_checks()` will return stable checks only by default. If `--include-experimental` is given, it will include experimental checks along with stable checks.
* `--categories` - Existing `Check_Categories::filter_checks_by_categories()` is used for filtering checks based on categories.
* `--stability` is removed and implemented `--include-experimental` flag.

**Review by felixarntz — COMMENTED** [review]
@ernilambar The production code changes look great to me now. However it looks like there are some Behat test failures, see https://github.com/WordPress/plugin-check/actions/runs/7470848675/job/20330167511?pr=379.

cc @swissspidy

**Comment by swissspidy** [comment]
Those Behat failures are unrelated and fixed in #381

**Review by felixarntz — APPROVED** [review]
Let's merge #381 and then refresh this branch to make sure the failures are gone :)

Thank you for the PR!

**Review by mukeshpanchal27 — APPROVED** [review]
